### PR TITLE
fix: Handle char expansion in islower, isupper

### LINKED ISSUE #7526: Disagreement with CPython for str.islower/isupper/swapcase
## Python Documentation or reference to CPython source code

tested on current main: 1a9b10ece50ebdbdd3820dafa8f6a630d87be9c0

There are many examples and I picked one.

```py
# the following is True in CPython and False in RustPython
print(
    "\u0295".islower() == True,
    "\u1c89".isupper() == False,
    "Σίσυφος".swapcase() == "σΊΣΥΦΟΣ",
)
```

### PR DESCRIPTION
**Pull request by joshuamegnauth54**
Closes: #7526

`py_islower` and `py_isupper` need to handle expansions for letter casing. Comparing chars directly can miss edge cases in certain languages. ~~Unfortunately, like the last PR, this allocates to handle potential expansions.~~

~~I also had to add `icu_casemap` as a dependency.~~
~~RustPython is already using parts of [icu4x](https://icu4x.unicode.org/) so this doesn't add many transitive dependencies.~~

**I figured out how to do it without allocation after working on CodeRabbit's suggestions.** I also added more tests based on CodeRabbit which all pass. I cross checked the tests with CPython too. :partying_face: 

---

The ICU crates use the Writeable trait internally which allows writing to a buffer. If you want to avoid the heap allocations, I can try wrapping a stack allocated buffer that falls back to heap allocations in my next patch. `icu_casemap` also handles title casing. It may be useful to drop `unicode-casing` in favor of `icu_casemap`. `icu4x` is the official Unicode consortium implementation, so it's robust and will be maintained indefinitely.

^ I wrote this comment earlier when I added `icu_casemap`. I still think it's relevant so I'll leave it. Currently, there are still title case bugs that may be easier to fix using `icu_casemap` over `unicode-casing`. I'll tackle that in another patch.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Unicode case detection and classification so string case checks (isupper/islower/istitle) behave correctly for a wider range of international characters and mixed emoji/text sequences.

* **Tests**
  * Expanded string tests to cover Unicode edge cases: modifier letters, title/swapcase behavior, emoji with text, and digit-only strings to ensure correct case-reporting.

### CODE REVIEW

**Comment by joshuamegnauth54** [comment]
Part of CI is failing because of two newly unused functions in the `AnyChar` trait. Should I just remove those functions?

File: extra_tests/snippets/builtin_str.py [hunk 1]
```diff
@@ -220,6 +222,14 @@
 assert "abc\t12345\txyz".expandtabs() == "abc     12345   xyz"
 assert "-".join(["1", "2", "3"]) == "1-2-3"
 assert "HALLO".isupper()
+assert not "123".isupper()
+assert not "123".islower()
+assert not "\U0001f431".isupper();
+assert not "\U0001f431".islower();
+assert "\U0001f431 CAT".isupper();
+assert "\U0001f431 cat".islower();
```
[the remark this answers is not in the document]
Comment (joshuamegnauth54) [reply extra_tests/snippets/builtin_str.py]: Oops. Brain is too rusted.

**Review by ShaharNaveh — APPROVED** [review]
lgtm:) 
my only concert is about the removal of those two methods from a public trait.

File: extra_tests/snippets/builtin_str.py [hunk 2]
```diff
@@ -220,6 +222,14 @@
...
 assert "-".join(["1", "2", "3"]) == "1-2-3"
 assert "HALLO".isupper()
+assert not "123".isupper()
+assert not "123".islower()
+assert not "\U0001f431".isupper()
+assert not "\U0001f431".islower()
+assert "\U0001f431 CAT".isupper()
+assert "\U0001f431 cat".islower()
+assert "\u0295".islower()
+assert "\u1c89".isupper()
```
Comment (ShaharNaveh) [inline extra_tests/snippets/builtin_str.py:232]: If CPython doesn't have a test case that cover those edge cases, I think it will be worth trying to get it upstream

**Comment by joshuamegnauth54** [comment]
Should I re-add the methods and mark them with `#[allow(unused)]`? CI failed so I wasn't sure what to do. :laughing:

**Review by youknowone — APPROVED** [review]
Thank you so much! They were removed because upper/lower case is not a trait of char, but trait of string, right?
